src/crypto.ts

```diff
@@ -0,0 +1,427 @@
+					(hashFunction) => {
+						// When the hash function resolves, create a Hasher and replace this object's methods with
+						// pointers to the corresponding methods on the Hasher.
+						const hasher = hashFunction.create(codec);
+						this.abort = bindOrUndefined(hasher, 'abort');
+						this.close = bindOrUndefined(hasher, 'close');
+						this.start = bindOrUndefined(hasher, 'start');
+						this.write = bindOrUndefined(hasher, 'write');
+						resolve(hasher);
+					},
+					function (error) {
+						reject(error);
+					}
+				);
+			})
+		});
+
+		Object.defineProperty(this, 'digest', {
+			enumerable: true,
+			value: new Promise((resolve, reject) => {
+				this._promise.then(
+					function (hasher) {
+						resolve(hasher.digest);
+					},
+					function (error) {
+						reject(error);
+					}
+				);
+			})
+		});
+	}
+
+	private _promise: Promise<Hasher<T>>;
+
+	digest: Promise<ByteBuffer>;
+
+	// Sink methods; the provider may or may not implement these, so call them with callOrNoop
+
+	abort(reason?: Error): Promise<void> {
+		return this._promise.then(function (hasher) {
+			return callOrNoop(hasher, 'abort', reason);
+		});
+	}
+
+	close(): Promise<void> {
+		return this._promise.then(function (hasher) {
+			return callOrNoop(hasher, 'close');
+		});
+	}
+
+	start(error: (error: Error) => void): Promise<void> {
+		return this._promise.then(function (hasher) {
+			return callOrNoop(hasher, 'start', error);
+		});
+	}
+
+	write(chunk: T): Promise<void> {
+		return this._promise.then(function (hasher) {
+			return callOrNoop(hasher, 'write', chunk);
+		});
+	}
+}
+
+/**
+ * A cryptographic key.
+ */
+export interface Key {
+	algorithm: string,
+	data: Data
+}
+
+/**
+ * An object for signing a data stream.
+ */
+export interface Signer<T extends Data> extends Sink<T> {
+	signature: Promise<ByteBuffer>;  // read only
+}
+
+/**
+ * A wrapper around a Promise<SignFunction> that will defer calls while a provider is asynchronously loaded.
+ */
+class SignerWrapper<T extends Data> implements Signer<T> {
+	constructor(signPromise: Promise<SignFunction>, key: Key, codec: Codec) {
+		Object.defineProperty(this, '_promise', {
+			value: new Promise((resolve, reject) => {
+				signPromise.then((signFunction) => {
+					// When the sign function resolves, create a Signer and replace this object's methods with
+					// pointers to the corresponding methods on the Signer.
+					const signer = signFunction.create(key, codec);
+					this.abort = bindOrUndefined(signer, 'abort');
+					this.close = bindOrUndefined(signer, 'close');
+					this.start = bindOrUndefined(signer, 'start');
+					this.write = bindOrUndefined(signer, 'write');
+					resolve(signer);
+				}).catch(function (error) {
+					reject(error);
+				});
+			})
+		});
+
+		Object.defineProperty(this, 'signature', {
+			value: new Promise((resolve, reject) => {
+				this._promise.then(
+					function (signer) {
+						resolve(signer.signature);
+					},
+					function (error) {
```

Comment:
this could just be `reject` instead of wrapped in an anonymous function. I understand if you want to leave it as is for consistency though